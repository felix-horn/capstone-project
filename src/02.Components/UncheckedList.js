import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd'
import ListItem from '../01.UI-Elements/ListItem'
// import DragHandleIcon from '@material-ui/icons/DragHandle'
import DragIndicatorIcon from '@material-ui/icons/DragIndicator'

UncheckedList.propTypes = {
  list: PropTypes.object.isRequired,
  listAllocation: PropTypes.array.isRequired,
  handleInputChange: PropTypes.func.isRequired,
  toggleIsChecked: PropTypes.func.isRequired,
  deleteListItem: PropTypes.func.isRequired,
  addListItemOnEnter: PropTypes.func.isRequired,
  checked: PropTypes.bool,
}

export default function UncheckedList({
  list,
  listAllocation,
  handleInputChange,
  toggleIsChecked,
  deleteListItem,
  addListItemOnEnter,
  checked,
}) {
  function handleOnDragEnd(result) {
    const indexFrom = result.source.index
    const indexTo = result.destination.index
    console.log({ indexFrom }, { indexTo })
  }

  return (
    <>
      <DragDropContext onDragEnd={handleOnDragEnd}>
        <Droppable droppableId="listItemsId">
          {(provided) => (
            <ListStyled
              checked={checked}
              {...provided.droppableProps}
              ref={provided.innerRef}
            >
              {listAllocation.map((id, index) => {
                const { title, isChecked } = list.byId[id]
                return (
                  <Draggable key={id} draggableId={id} index={index}>
                    {(provided) => (
                      <Wrapper
                        {...provided.draggableProps}
                        {...provided.dragHandleProps}
                        ref={provided.innerRef}
                      >
                        <DragIndicatorIconStyled />
                        {/* TEXT */}
                        <ListItem
                          id={id}
                          title={title}
                          isChecked={isChecked}
                          onInputChange={handleInputChange}
                          onToggleCheckbox={() => toggleIsChecked(id)}
                          onDelete={() => deleteListItem(id)}
                          onEnter={() => addListItemOnEnter(id)}
                        />
                      </Wrapper>
                    )}
                  </Draggable>
                )
              })}
              {provided.placeholder}
            </ListStyled>
          )}
        </Droppable>
      </DragDropContext>
    </>
  )
}

const Wrapper = styled.div`
  display: flex;
  align-items: center;
`

const DragIndicatorIconStyled = styled(DragIndicatorIcon)`
  opacity: 0.3;
`

const ListStyled = styled.div`
  margin: 3px 2px 10px;
  padding-top: ${(props) => (props.checked ? '10px' : 'none')};
  border-top: ${(props) => (props.checked ? '1px solid #dadcdf' : 'none')};
  display: grid;
`
