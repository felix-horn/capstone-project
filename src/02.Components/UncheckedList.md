```js padded
<UncheckedList 
    
    list={{
        allIds: ["a","b"],
        byId: {
            "a": { id: "a", title: "Koks", isChecked: false },
            "b": { id: "b", title: "Nutten", isChecked: false},
        }
    }}
/>
```