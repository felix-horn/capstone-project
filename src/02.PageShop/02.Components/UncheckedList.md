```js padded
<UncheckedList 
shopId={'x'} 
database={{
    shops: {
    allIds: ['x'],
    byId: {
      x: {
        id: 'x',
        title: 'Shop X',
        items: ['a', 'b'],
      },
    },
  },
  items: {
    allIds: ['a', 'b'],
    byId: {
      a: {
        id: 'a',
        title: 'Item A',
        isChecked: false,
      },
      b: {
        id: 'b',
        title: 'Item B',
        isChecked: false,
      },      
    },
  },
}}
/>
```