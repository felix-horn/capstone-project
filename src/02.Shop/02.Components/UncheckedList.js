import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd'
import ListItem from '../01.UI-Elements/ListItem'
import DragIndicatorIcon from '@material-ui/icons/DragIndicator'
import {
  getShopItemIndex,
  getUncheckedItemIds,
} from '../../services/filter.services'

UncheckedList.propTypes = {
  shopId: PropTypes.string.isRequired,
  database: PropTypes.object.isRequired,
  changeTitle: PropTypes.func.isRequired,
  toggleIsChecked: PropTypes.func.isRequired,
  deleteListItem: PropTypes.func.isRequired,
  addListItem: PropTypes.func.isRequired,
  rearrangeListOrder: PropTypes.func.isRequired,
  checked: PropTypes.bool,
}

export default function UncheckedList({
  shopId,
  database,
  changeTitle,
  toggleIsChecked,
  deleteListItem,
  addListItem,
  rearrangeListOrder,
}) {
  const uncheckedIds = getUncheckedItemIds(database, shopId)

  return (
    <DragDropContext
      onDragStart={giveHapticFeedback}
      onDragEnd={handleOnDragEnd}
    >
      <Droppable droppableId="listId">
        {(provided) => (
          <UncheckedListLayout
            {...provided.droppableProps}
            ref={provided.innerRef}
          >
            {uncheckedIds.map((id, index) => (
              <Draggable key={id} draggableId={id} index={index}>
                {(provided, snapshot) => (
                  <ListItemWrapper
                    {...provided.draggableProps}
                    ref={provided.innerRef}
                    isDragging={snapshot.isDragging}
                  >
                    <DragIconWrapper {...provided.dragHandleProps}>
                      <DragIndicatorIcon />
                    </DragIconWrapper>
                    <ListItem
                      itemId={id}
                      shopId={shopId}
                      database={database}
                      changeTitle={(fieldValue) => changeTitle(id, fieldValue)}
                      toggleCheckbox={() => toggleIsChecked(id)}
                      onDelete={() => deleteListItem(id)}
                      onEnter={() => handleEnter(id)}
                    />
                  </ListItemWrapper>
                )}
              </Draggable>
            ))}
            {provided.placeholder}
          </UncheckedListLayout>
        )}
      </Droppable>
    </DragDropContext>
  )
  function giveHapticFeedback() {
    if (window.navigator.vibrate) {
      window.navigator.vibrate(10)
    }
  }

  function handleOnDragEnd(result) {
    // return early if item is dragged outside droppable area
    if (!result.destination) return
    const draggedId = result.draggableId
    // new index within unchecked IDs is returned
    const newIndexInUncheckedIds = result.destination.index
    // new index within all IDs of the shop (including checked IDs) must be determined
    const newIdPrecedingDraggedId = uncheckedIds[newIndexInUncheckedIds - 1]
    const newIndexInShopIds =
      getShopItemIndex(database, shopId, newIdPrecedingDraggedId) + 1
    rearrangeListOrder(draggedId, newIndexInShopIds)
  }

  function handleEnter(targetId) {
    // only for the last item of the unchecked list enter creates new item
    const lastUncheckedId = uncheckedIds[uncheckedIds.length - 1]
    lastUncheckedId === targetId && addListItem()
  }
}

const ListItemWrapper = styled.div`
  box-shadow: ${(props) =>
    props.isDragging ? 'var(--light-box-shadow)' : 'none'};
  border-radius: 5px;
  background: var(--white);
  display: flex;
  align-items: center;
`

const DragIconWrapper = styled.div`
  display: flex;
  align-items: center;
  opacity: 0.3;
`

const UncheckedListLayout = styled.div`
  margin: 3px 2px 10px;
  display: grid;
`
