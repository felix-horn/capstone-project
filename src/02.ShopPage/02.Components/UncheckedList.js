import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd'
import ListItem from '../01.UI-Elements/ListItem'
import DragIndicatorIcon from '@material-ui/icons/DragIndicator'

UncheckedList.propTypes = {
  database: PropTypes.object.isRequired,
  changeTitle: PropTypes.func.isRequired,
  toggleIsChecked: PropTypes.func.isRequired,
  deleteListItem: PropTypes.func.isRequired,
  addListItem: PropTypes.func.isRequired,
  rearrangeListOrder: PropTypes.func.isRequired,
  checked: PropTypes.bool,
}

export default function UncheckedList({
  shopId,
  database,
  changeTitle,
  toggleIsChecked,
  deleteListItem,
  addListItem,
  rearrangeListOrder,
}) {
  /* const uncheckedIds = database.items.allIds.filter(
    (id) => !database.items.byId[id].isChecked
  ) */

  console.log(shopId)
  console.log(database.shops.byId[shopId].items)

  const uncheckedIds = database.shops.byId[shopId].items.filter(
    (id) => !database.items.byId[id].isChecked
  )

  console.log(uncheckedIds)
  return (
    <>
      <DragDropContext
        onDragStart={giveHapticFeedback}
        onDragEnd={handleOnDragEnd}
      >
        <Droppable droppableId="listId">
          {(provided) => (
            <UncheckedListStyled
              {...provided.droppableProps}
              ref={provided.innerRef}
            >
              {uncheckedIds.map((id, index) => {
                const { title, isChecked } = database.items.byId[id] //check whether isChecked is obsolete
                return (
                  <Draggable key={id} draggableId={id} index={index}>
                    {(provided, snapshot) => (
                      <ListItemWrapper
                        {...provided.draggableProps}
                        ref={provided.innerRef}
                        isDragging={snapshot.isDragging}
                      >
                        <DragIconWrapper {...provided.dragHandleProps}>
                          <DragIndicatorIcon />
                        </DragIconWrapper>
                        <ListItem
                          id={id}
                          title={title}
                          isChecked={isChecked}
                          changeTitle={(fieldValue) =>
                            changeTitle(id, fieldValue)
                          }
                          toggleCheckbox={() => toggleIsChecked(id)}
                          onDelete={() => deleteListItem(id)}
                          onEnter={() => handleEnter(id)}
                        />
                      </ListItemWrapper>
                    )}
                  </Draggable>
                )
              })}
              {provided.placeholder}
            </UncheckedListStyled>
          )}
        </Droppable>
      </DragDropContext>
    </>
  )
  function giveHapticFeedback() {
    if (window.navigator.vibrate) {
      window.navigator.vibrate(10)
    }
  }

  function handleOnDragEnd(result) {
    if (!result.destination) return
    const indexFrom = result.source.index
    const indexTo = result.destination.index
    rearrangeListOrder(indexFrom, indexTo)
  }

  function handleEnter(targetId) {
    const lastUncheckedId = uncheckedIds[uncheckedIds.length - 1]
    lastUncheckedId === targetId && addListItem()
  }
}

const ListItemWrapper = styled.div`
  box-shadow: ${(props) =>
    props.isDragging ? 'var(--light-box-shadow)' : 'none'};
  border-radius: 5px;
  background: var(--white);
  display: flex;
  align-items: center;
`

const DragIconWrapper = styled.div`
  display: flex;
  align-items: center;
  opacity: 0.3;
`

const UncheckedListStyled = styled.div`
  margin: 3px 2px 10px;
  display: grid;
`
