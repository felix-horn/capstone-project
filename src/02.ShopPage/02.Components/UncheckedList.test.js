import { render } from '@testing-library/react'
import UncheckedList from './UncheckedList'

const testDatabase = {
  shops: {
    allIds: ['x', 'y', 'z'],
    byId: {
      x: {
        id: 'x',
        title: 'Penny',
        items: ['a', 'b'],
      },
      y: {
        id: 'y',
        title: 'Budni',
        items: ['c'],
      },
    },
  },
  items: {
    allIds: ['a', 'b', 'c'],
    byId: {
      a: {
        id: 'a',
        title: 'Milch',
        isChecked: false,
      },
      b: {
        id: 'b',
        title: 'Butter',
        isChecked: true,
      },
      c: {
        id: 'c',
        title: 'Shampoo',
        isChecked: false,
      },
    },
  },
}

describe('UncheckedList', () => {
  it('renders the list correctly', () => {
    const { container } = render(<UncheckedList database={testDatabase} />)
    expect(container.firstChild).toMatchSnapshot()
  })
  fit('only renders allocated listItems', () => {
    const { queryByDisplayValue, getByDisplayValue } = render(
      <UncheckedList database={testDatabase} />
    )
    const milkListItem = getByDisplayValue(/milch/i)
    const cheeseListItem = queryByDisplayValue(/butter/i)
    expect(milkListItem).toBeInTheDocument()
    expect(cheeseListItem).not.toBeInTheDocument()
  })
})
