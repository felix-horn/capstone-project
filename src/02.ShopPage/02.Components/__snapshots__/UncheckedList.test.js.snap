// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UncheckedList renders the list correctly 1`] = `
.c3 {
  width: 75vw;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  opacity: 1;
}

.c4 {
  margin-left: 9px;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: none;
}

.c4:focus {
  outline: none;
}

.c5 {
  margin-left: auto;
  -webkit-transform: scale(0.9);
  -ms-transform: scale(0.9);
  transform: scale(0.9);
  opacity: 0.4;
}

.c1 {
  box-shadow: none;
  border-radius: 5px;
  background: var(--white);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  opacity: 0.3;
}

.c0 {
  margin: 3px 2px 10px;
  display: grid;
}

<div
  class="c0"
  data-rbd-droppable-context-id="0"
  data-rbd-droppable-id="listId"
>
  <div
    class="c1"
    data-rbd-draggable-context-id="0"
    data-rbd-draggable-id="a"
  >
    <div
      aria-describedby="rbd-hidden-text-0-hidden-text-0"
      class="c2"
      data-rbd-drag-handle-context-id="0"
      data-rbd-drag-handle-draggable-id="a"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm-2-8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </div>
    <label
      class="c3"
      data-testid="list-item"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root"
        data-testid="checkbox"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-4"
            data-indeterminate="false"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <input
        class="c4"
        data-testid="title-list-item"
        value="Milch"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root c5"
        data-testid="delete-list-item"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </label>
  </div>
  <div
    class="c1"
    data-rbd-draggable-context-id="0"
    data-rbd-draggable-id="b"
  >
    <div
      aria-describedby="rbd-hidden-text-0-hidden-text-0"
      class="c2"
      data-rbd-drag-handle-context-id="0"
      data-rbd-drag-handle-draggable-id="b"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm-2-8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </div>
    <label
      class="c3"
      data-testid="list-item"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root"
        data-testid="checkbox"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-4"
            data-indeterminate="false"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <input
        class="c4"
        data-testid="title-list-item"
        value="Butter"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root c5"
        data-testid="delete-list-item"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </label>
  </div>
  <div
    class="c1"
    data-rbd-draggable-context-id="0"
    data-rbd-draggable-id="c"
  >
    <div
      aria-describedby="rbd-hidden-text-0-hidden-text-0"
      class="c2"
      data-rbd-drag-handle-context-id="0"
      data-rbd-drag-handle-draggable-id="c"
      draggable="false"
      role="button"
      tabindex="0"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm-2-8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </div>
    <label
      class="c3"
      data-testid="list-item"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiCheckbox-root"
        data-testid="checkbox"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-4"
            data-indeterminate="false"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <input
        class="c4"
        data-testid="title-list-item"
        value="Shampoo"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root c5"
        data-testid="delete-list-item"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </label>
  </div>
</div>
`;
